Updating a sub-attribute of a template in `mso_schema` fails when `template_type` is not specified in the Terraform config.

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version

Terraform v1.5.3
on darwin_arm64
+ provider registry.terraform.io/ciscodevnet/mso v0.11.0

### MSO version

* NDO V 4.1.2e

### Affected Resource(s)

* mso_schema

### Expected Behavior

When a `description` attribute is added to an existing template in a `mso_schema` resource, the description will be updated.

### Actual Behavior

The provider attempts to remove and create a replacement template block.

Note: This also applies when adding new template blocks to an existing `mso_schema` resource.

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. Define the schema with one template and no description:

```hcl
resource "mso_schema" "schema" {
    name = "Bug"

    template {
        name          = "Template1"
        display_name  = "Template1"
        tenant_id     = "64b8e6f789c60aeac323c7c5"
    }
}
```

2. `terraform apply`

3. Update the template to include a description:

```hcl
resource "mso_schema" "schema" {
    name = "Bug"

    template {
        name          = "Template1"
        display_name  = "Template1"
        tenant_id     = "64b8e6f789c60aeac323c7c5"
        description   = "bug"
    }
}
```

4. `terraform plan`

```
  # module.ndo.mso_schema.schema will be updated in-place
  ~ resource "mso_schema" "schema" {
        id   = "64b8e6fa89c60aeac323c7c8"
        name = "Bug"

      - template {
          - display_name  = "Template1" -> null
          - name          = "Template1" -> null
          - template_type = "aci_multi_site" -> null
          - tenant_id     = "64b8e6f789c60aeac323c7c5" -> null
        }
      + template {
          + description   = "bug"
          + display_name  = "Template1"
          + name          = "Template1"
          + template_type = (known after apply)
          + tenant_id     = "64b8e6f789c60aeac323c7c5"
        }
    }
```

Here is sanitized debug output when it fails to apply the above change:

```
2023-07-24T09:30:38.716+1000 [WARN]  Provider "registry.terraform.io/ciscodevnet/mso" produced an invalid plan for module.ndo.mso_schema.schema["Bug"], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .template_name: planned value cty.StringVal("") for a non-computed attribute
      - .tenant_id: planned value cty.StringVal("") for a non-computed attribute
module.ndo.mso_schema.schema["Bug"]: Modifying... [id=64b8e6fa89c60aeac323c7c8]
2023-07-24T09:30:38.716+1000 [INFO]  Starting apply for module.ndo.mso_schema.schema["Bug"]
2023-07-24T09:30:38.717+1000 [DEBUG] module.ndo.mso_schema.schema["Bug"]: applying the planned Update change
2023-07-24T09:30:38.717+1000 [DEBUG] provider.terraform-provider-mso_v0.11.0: 2023/07/24 09:30:38 [DEBUG] Schema: Beginning Update
2023-07-24T09:30:38.928+1000 [ERROR] vertex "module.ndo.mso_schema.schema[\"Bug\"]" error: Template type cannot be changed.
2023-07-24T09:30:38.928+1000 [DEBUG] cloud/state: state read serial is: 44; serial is: 44
2023-07-24T09:30:38.928+1000 [DEBUG] cloud/state: state read lineage is: 8e38bbc5-7cfa-e8e6-1e38-b8bdb83a3613; lineage is: 8e38bbc5-7cfa-e8e6-1e38-b8bdb83a3613
╷
│ Error: Template type cannot be changed.
│ 
│   with module.ndo.mso_schema.schema["Bug"],
│   on ../../../terraform-mso-nac-ndo/ndo_schemas.tf line 58, in resource "mso_schema" "schema":
│   58: resource "mso_schema" "schema" {
│ 
╵
2023-07-24T09:30:40.499+1000 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-07-24T09:30:40.502+1000 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/ciscodevnet/mso/0.11.0/darwin_arm64/terraform-provider-mso_v0.11.0 pid=10470
2023-07-24T09:30:40.502+1000 [DEBUG] provider: plugin exited
```

If I initially create the template with a description and then try to change it, the same issue occurs. However if I add a second template it works fine as long as the existing template has a description set.

There appears to be an issue with the way the description attribute was implemented in https://github.com/CiscoDevNet/terraform-provider-mso/commit/b9c7f0c33d5f094ca316819d81b4339ee14d7b99.

--- a/mso/resource_mso_schema.go
+++ b/mso/resource_mso_schema.go
@@ -88,6 +88,21 @@ func resourceMSOSchema() *schema.Resource {
 				},
 			},
 		}),
+		CustomizeDiff: func(diff *schema.ResourceDiff, v interface{}) error {
+			// check if template_type is changed between known state and provided configuration and error out during plan if it is
+			stateTemplate, configTemplate := diff.GetChange("template")
+			for _, valueState := range stateTemplate.(*schema.Set).List() {
+				state := valueState.(map[string]interface{})
+				for _, valueConfig := range configTemplate.(*schema.Set).List() {
+					config := valueConfig.(map[string]interface{})
+					if state["name"] == config["name"] && state["template_type"] != config["template_type"] &&
+						!(state["template_type"] == "aci_multi_site" && config["template_type"] == "") {
+						return fmt.Errorf("Template type cannot be changed. Change detected from '%s' to '%s'.", state["template_type"], config["template_type"])
+					}
+				}
+			}
+			return nil
+		},
 	}
 }
 
@@ -184,8 +199,10 @@ func resourceMSOSchemaCreate(d *schema.ResourceData, m interface{}) error {
 				map_templates["displayName"] = inner_templates["display_name"]
 				map_templates["tenantId"] = inner_templates["tenant_id"]
 				map_templates["description"] = inner_templates["description"]
-				map_templates["templateType"] = getTemplateType(inner_templates["template_type"].(string))
-				map_templates["templateSubType"] = getTemplateSubType(inner_templates["template_type"].(string))
+				if val, ok := inner_templates["template_type"]; ok && val.(string) != "" {
+					map_templates["templateType"] = getTemplateType(inner_templates["template_type"].(string))
+					map_templates["templateSubType"] = getTemplateSubType(inner_templates["template_type"].(string))
+				}
 				templates = append(templates, map_templates)
 			}
 		}
@@ -237,7 +254,9 @@ func resourceMSOSchemaImport(d *schema.ResourceData, m interface{}) ([]*schema.R
 		map_template["display_name"] = models.StripQuotes(templatesCont.S("displayName").String())
 		map_template["tenant_id"] = models.StripQuotes(templatesCont.S("tenantId").String())
 		map_template["description"] = models.StripQuotes(templatesCont.S("description").String())
-		map_template["template_type"] = getSchemaTemplateType(templatesCont)
+		if templatesCont.Exists("templateType") {
+			map_template["template_type"] = getSchemaTemplateType(templatesCont)
+		}
 		templates = append(templates, map_template)
 
 	}
@@ -374,9 +393,6 @@ func resourceMSOSchemaUpdate(d *schema.ResourceData, m interface{}) error {
 				valueOld := valueMapOld.(map[string]interface{})
 				for _, valueMapNew := range getDifferenceNew {
 					valueNew := valueMapNew.(map[string]interface{})
-					if valueOld["name"] == valueNew["name"] && valueOld["template_type"] != valueNew["template_type"] {
-						return fmt.Errorf("Template type cannot be changed.")
-					}
 
 					// Tenant Id of template has been changed
 					if valueOld["name"] == valueNew["name"] && valueOld["tenant_id"] != valueNew["tenant_id"] {
@@ -435,13 +451,11 @@ func resourceMSOSchemaUpdate(d *schema.ResourceData, m interface{}) error {
 			for _, MapToAdd := range listMapsToAdd {
 
 				changedMap := MapToAdd.(map[string]interface{})
-				templateType := getTemplateType(changedMap["template_type"].(string))
-				templateSubType := getTemplateSubType(changedMap["template_type"].(string))
-				if _, ok := changedMap["template_type"]; ok {
-					delete(changedMap, "template_type")
+
+				if val, ok := changedMap["template_type"]; ok && val.(string) != "" {
+					changedMap["templateType"] = getTemplateType(changedMap["template_type"].(string))
+					changedMap["templateSubType"] = getTemplateSubType(changedMap["template_type"].(string))
 				}
-				changedMap["templateType"] = templateType
-				changedMap["templateSubType"] = templateSubType
 
 				map_add, _ := json.Marshal(changedMap)
 				map_values := strings.Replace(strings.Replace(string(map_add), "display_name", "displayName", 1), "tenant_id", "tenantId", 1)
@@ -541,7 +555,9 @@ func resourceMSOSchemaRead(d *schema.ResourceData, m interface{}) error {
 		map_template["display_name"] = models.StripQuotes(templatesCont.S("displayName").String())
 		map_template["tenant_id"] = models.StripQuotes(templatesCont.S("tenantId").String())
 		map_template["description"] = models.StripQuotes(templatesCont.S("description").String())
-		map_template["template_type"] = getSchemaTemplateType(templatesCont)
+		if templatesCont.Exists("templateType") {
+			map_template["template_type"] = getSchemaTemplateType(templatesCont)
+		}
 		templates = append(templates, map_template)
 
 		apiTemplate := models.StripQuotes(templatesCont.S("name").String())

--- a/mso/resource_mso_schema.go
+++ b/mso/resource_mso_schema.go
@@ -80,7 +80,7 @@ func resourceMSOSchema() *schema.Resource {
 						"template_type": &schema.Schema{
 							Type:     schema.TypeString,
 							Optional: true,
-							Computed: true,
+							Default:  "aci_multi_site",
 							// validation func does not work inside typeset
 							ValidateFunc: validation.StringInSlice(getSchemaTemplateTypes(), false),
 						},

--- a/website/docs/r/schema.html.markdown
+++ b/website/docs/r/schema.html.markdown
@@ -51,7 +51,7 @@ resource "mso_schema" "demo_schema" {
   * `display_name` - (Required) The display name of the template.
   * `tenant_id` - (Required) The tenant-id to associate with the template.
   * `description` - (Optional) The description of the template.
-  * `template_type` - (Optional) The template type of the template. Allowed values are `aci_multi_site`, `aci_autonomous`, `ndfc`, `cloud_local`, and `sr_mpls`. NDO defaults to `aci_multi_site` when attribute is unset during creation.
+  * `template_type` - (Optional) The template type of the template. Allowed values are `aci_multi_site`, `aci_autonomous`, `ndfc`, `cloud_local`, and `sr_mpls`. Defaults to `aci_multi_site` when attribute is unset during creation.
 
 ## Attribute Reference ##
 

